modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java

```diff
@@ -917,13 +927,63 @@ else if (type == EVT_CLIENT_NODE_RECONNECTED) {
                             discoCache,
                             notification.getTopSnapshot(),
                             stateFinishMsg,
-                            notification.getSpanContainer()
+                            notification.getSpanContainer(),
+                            secCtx
                         )
                     );
 
                 if (type == EVT_CLIENT_NODE_DISCONNECTED)
                     discoWrk.awaitDisconnectEvent();
             }
+
+            /** */
+            class SecurityAwareNotificationTask extends NotificationTask {
+                /** */
+                public SecurityAwareNotificationTask(DiscoveryNotification notification) {
+                    super(notification);
+                }
+
+                /** */
+                @Override public void run() {
+                    DiscoverySpiCustomMessage customMsg = notification.getCustomMsgData();
+
+                    if (customMsg instanceof SecurityAwareCustomMessageWrapper) {
+                        UUID secSubjId = ((SecurityAwareCustomMessageWrapper)customMsg).securitySubjectId();
+
+                        try (OperationSecurityContext ignored = ctx.security().withContext(secSubjId)) {
+                            super.run();
+                        }
+                    }
+                    else {
+                        SecurityContext initiatorNodeSecCtx = nodeSecurityContext(
+                            marshaller,
+                            U.resolveClassLoader(ctx.config()),
+                            notification.getNode()
+                        );
+
+                        try (OperationSecurityContext ignored = ctx.security().withContext(initiatorNodeSecCtx)) {
```

Comment:
The main problem in this case is that IgniteSecurityProcessor#withContext(UUID) method uses GridDiscoveryManager#node(UUID) to obtain a node instance. But DiscoveryCache is updated DURING event handling (see `topSnap.set(new Snapshot(snapshot.topVer, discoCache));` calls in DiscoverySpiListener#onDiscovery0) but we must get node security context BEFORE event handling. So in case of NODE_JOINED event we will see NPE during IgniteSecurityProcessor#withContext(UUID) call as no node with specified ID will be found.
Since internal discovery events such as NODE_JOINED or NODE_LEFT does not occur so often, I think that we can sacrifice caching. WDYT?

Reply:
Ok, agree